ext/windows/service/daemon.rb

```diff
@@ -54,25 +54,33 @@ def service_main(*argsv)
     #   )
     # end
 
-    puppet = File.join(basedir, 'bin', 'puppet.bat')
+    puppet = File.join(basedir, 'puppet', 'bin', 'puppet')
+    ruby = File.join(basedir, 'puppet', 'bin', 'ruby.exe')
+    ruby_puppet_cmd = "\"#{ruby}\" \"#{puppet}\""
+
     unless File.exist?(puppet)
       log_err("File not found: '#{puppet}'")
       return
     end
     log_debug("Using '#{puppet}'")
 
     cmdline_debug = argsv.index('--debug') ? :debug : nil
-    @loglevel = parse_log_level(puppet, cmdline_debug)
+    @loglevel = parse_log_level(ruby_puppet_cmd, cmdline_debug)
     log_notice('Service started')
 
     service = self
     @run_thread = Thread.new do
       begin
         while service.running? do
-          runinterval = service.parse_runinterval(puppet)
+          runinterval = service.parse_runinterval(ruby_puppet_cmd)
           if service.state == RUNNING or service.state == IDLE
+            ENV['SSL_CERT_FILE'] = File.join(basedir, 'puppet', 'ssl', 'cert.pem')
+            ENV['SSL_CERT_DIR'] = File.join(basedir, 'puppet', 'ssl', 'certs')
+            ENV['OPENSSL_CONF'] = File.join(basedir, 'puppet' ,'ssl', 'openssl.cnf')
+            ENV['FACTER_env_windows_installdir'] = basedir
```

Comment:
This isn't quite right. The `environment.bat` results in the following changes to the current environment:

```diff
11a12,13
> FACTERDIR=C:\Program Files\Puppet Labs\Puppet\facter
> FACTER_env_windows_installdir=C:\Program Files\Puppet Labs\Puppet
12a15
> HIERA_DIR=C:\Program Files\Puppet Labs\Puppet\hiera
16a20
> MCOLLECTIVE_DIR=C:\Program Files\Puppet Labs\Puppet\mcollective
17a22
> OPENSSL_CONF=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\openssl.cnf
19c24
< Path=C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Packer\SysInternals;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\6\;C:\Program Files\Puppet Labs\Puppet\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;
---
> Path=C:\Program Files\Puppet Labs\Puppet\puppet\bin;C:\Program Files\Puppet Labs\Puppet\facter\bin;C:\Program Files\Puppet Labs\Puppet\hiera\bin;C:\Program Files\Puppet Labs\Puppet\mcollective\bin;C:\Program Files\Puppet Labs\Puppet\bin;C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin;C:\Program Files\Puppet Labs\Puppet\sys\tools\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Packer\SysInternals;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\6\;C:\Program Files\Puppet Labs\Puppet\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;
20a26
> PL_BASEDIR=C:\Program Files\Puppet Labs\Puppet
31a38,41
> PUPPET_DIR=C:\Program Files\Puppet Labs\Puppet\puppet
> RUBYLIB=C:/Program Files/Puppet Labs/Puppet/puppet/lib;C:/Program Files/Puppet Labs/Puppet/facter/lib;C:/Program Files/Puppet Labs/Puppet/hiera/lib;C:/Program Files/Puppet Labs/Puppet/mcollective/lib;
> RUBYOPT=rubygems
> RUBY_DIR=C:\Program Files\Puppet Labs\Puppet\sys\ruby
32a43,44
> SSL_CERT_DIR=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\certs
> SSL_CERT_FILE=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\cert.pem
```

Comparing what `environment.bat` current does to `daemon.rb` in this PR I get:

```diff
1,13c1,6
< FACTERDIR=C:\Program Files\Puppet Labs\Puppet\facter
< FACTER_env_windows_installdir=C:\Program Files\Puppet Labs\Puppet
< HIERA_DIR=C:\Program Files\Puppet Labs\Puppet\hiera
< MCOLLECTIVE_DIR=C:\Program Files\Puppet Labs\Puppet\mcollective
< OPENSSL_CONF=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\openssl.cnf
< Path=C:\Program Files\Puppet Labs\Puppet\puppet\bin;C:\Program Files\Puppet Labs\Puppet\facter\bin;C:\Program Files\Puppet Labs\Puppet\hiera\bin;C:\Program Files\Puppet Labs\Puppet\mcollective\bin;C:\Program Files\Puppet Labs\Puppet\bin;C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin;C:\Program Files\Puppet Labs\Puppet\sys\tools\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Packer\SysInternals;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\6\;C:\Program Files\Puppet Labs\Puppet\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;
< PL_BASEDIR=C:\Program Files\Puppet Labs\Puppet
< PUPPET_DIR=C:\Program Files\Puppet Labs\Puppet\puppet
< RUBYLIB=C:/Program Files/Puppet Labs/Puppet/puppet/lib;C:/Program Files/Puppet Labs/Puppet/facter/lib;C:/Program Files/Puppet Labs/Puppet/hiera/lib;C:/Program Files/Puppet Labs/Puppet/mcollective/lib;
< RUBYOPT=rubygems
< RUBY_DIR=C:\Program Files\Puppet Labs\Puppet\sys\ruby
< SSL_CERT_DIR=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\certs
< SSL_CERT_FILE=C:\Program Files\Puppet Labs\Puppet\puppet\ssl\cert.pem
---
> FACTER_env_windows_installdir=C:/Program Files/Puppet Labs/Puppet
> OPENSSL_CONF=C:/Program Files/Puppet Labs/Puppet/puppet/ssl/openssl.cnf
> Path=C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Packer\SysInternals;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\6\;C:\Program Files\Puppet Labs\Puppet\bin;C:\Windows\system32\config\systemprofile\AppData\Local\Microsoft\WindowsApps
> SSL_CERT_DIR=C:/Program Files/Puppet Labs/Puppet/puppet/ssl/certs
> SSL_CERT_FILE=C:/Program Files/Puppet Labs/Puppet/puppet/ssl/cert.pem
>
```

1. `PL_BASEDIR=C:\Program Files\Puppet Labs\Puppet` is missing
1. The env variable for each component is missing, eg `FACTERDIR`, `HIERA_DIR`, `MCOLLECTIVE_DIR`, `PUPPET_DIR`. I think we changed `FACTERDIR` to be different, as it used to define `DIR` as a fact.
1. The ruby component is missing and is special, `RUBY_DIR=C:\Program Files\Puppet Labs\Puppet\sys\ruby`
1. Each components' bin directory needs to be **prepended** to `PATH` using backslashes
1. Each components's lib directory needs to be set in `RUBYLIB` using forward slashes
1. We set `RUBYOPT=rubygems` to force rubygems to be loaded. I have no idea if this is still required, but we should keep that.
1. `PATH` is also missing:
* `C:\Program Files\Puppet Labs\Puppet\bin` - It shouldn't be required anymore, since we're no longer calling the `.bat` files, but we should preserve it just in case.
* `C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin` - This is necessary so that ruby (and friends) are in the path. We could also force have `daemon.rb` set `ruby = "C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin\ruby.exe"` so that we're not relying on the search path.
* `C:\Program Files\Puppet Labs\Puppet\sys\tools\bin` - This contains `elevate` which isn't needed by the daemon, but we should include it for now.
1. OpenSSL environment variables should use backslashes.

Reply:
Yes, confirmed that if some of this are not exported, also the ruby bin folder is different in puppet 5 and puppet 6. I'm trying to find where would best place to set the environment in puppet. @joshcooper any suggestions on this?

Reply:
As part of path restructuring for Puppet6, a number of superfluous environment variables were removed from `environment.bat` in https://github.com/puppetlabs/puppet-agent/commit/10df00fca2034c353cf543086d263fd8490f76b4#diff-f19be7ccad3d10336f31489de24d09dd by @caseywilliams (and IIRC @McdonaldSeanp 	was involved as well)

I can understand wanting to keep things identical to what they were previously, but a number of these changes are already made in Puppet 6 without any discernible impact (granted, on a major version boundary). So at the very least the changes here will be different when merged up to the Puppet 6 branch. 

Most environment variable values were only ever an implementation detail inside the batch file, weren't intended to be consumed as environment variables, and were never documented as values that were usable by end users. We really should have used the batch file `SETLOCAL` command to make the scoping more clear, but the original authors of `environment.bat` might not have been familiar with some of the esoteric batch file rules. With that said, I don't think the following environment variables are important to define:

* `PL_BASEDIR`
* `PUPPET_DIR`
* `HIERA_DIR`
* `RUBY_DIR`

If there were any consumers of these leaked variables (i.e. module content), I think we would have seen bug reports in JIRA. I searched JIRA and didn't find anything.

So to go down the list of issues @joshcooper raised:

> 1. `PL_BASEDIR=C:\Program Files\Puppet Labs\Puppet` is missing

Unnecessary. 

This is why the `env_windows_installdir` Fact exists (with the same value), is [documented for end users to use](https://puppet.com/docs/facter/3.3/core_facts.html#envwindowsinstalldir), and is a value that other Puppet code uses.

> 2. The env variable for each component is missing, eg `FACTERDIR`, `HIERA_DIR`, `MCOLLECTIVE_DIR`, `PUPPET_DIR`. I think we changed `FACTERDIR` to be different, as it used to define `DIR` as a fact.

Aside from `FACTERDIR` (which should be defined), these were only intended to be variables inside the batch file, and should be removable.

`FACTERDIR` helps locate `libfacter.so` (really a `.dll`), but I think that in Puppet6 it was removed because `ruby.exe` and `libfacter.so` are now both in the same directory, but they're not in Puppet 5.

`FACTERDIR` is also used in the Facter Java shim at https://github.com/puppetlabs/facter/blob/dcd15b277db3773e0a9a1efeff8a4f0272a7b4dc/lib/Facter.java.in#L12 but that isn't relevant to Windows

> 3. The ruby component is missing and is special, `RUBY_DIR=C:\Program Files\Puppet Labs\Puppet\sys\ruby`
> 4. Each components' bin directory needs to be **prepended** to `PATH` using backslashes

`RUBY_DIR` was never used anywhere outside of `environment.bat`.

It *is* important, specifically for the Puppet 5 case to make sure that `PATH` includes the additional paths to binaries as it did previously like:

```
SET PATH=%PUPPET_DIR%\bin;%FACTERDIR%\bin;%HIERA_DIR%\bin;%PL_BASEDIR%\bin;%RUBY_DIR%\bin;%PL_BASEDIR%\sys\tools\bin;%PATH%
```

Note that doing this in Ruby code may actually fix a bug in how `environment.bat` combines paths - see https://tickets.puppetlabs.com/browse/PA-2156

> 5. Each components lib directory needs to be set in `RUBYLIB` using forward slashes

Agreed. 

`RUBYLIB` can be set as it was before as an environment variable: 

```
SET RUBYLIB=%PUPPET_DIR%\lib;%FACTERDIR%\lib;%HIERA_DIR%\lib;%RUBYLIB%
```

Alternatively, we can pass `-I` with these paths to `ruby.exe`, instead of setting an environment variable.


> 6. We set `RUBYOPT=rubygems` to force rubygems to be loaded. I have no idea if this is still required, but we should keep that.

This is not necessary, because Rubygems has been part of Ruby since 1.9 (no call to `require` would work without `rubygems` already being loaded). This is documented at https://guides.rubygems.org/rubygems-basics/#requiring-code

> 7. OpenSSL environment variables should use backslashes.

Agreed

> 8. `PATH` is also missing:
> * `C:\Program Files\Puppet Labs\Puppet\bin` - It shouldn't be required anymore, since we're no longer calling the `.bat` files, but we should preserve it just in case.
> * `C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin` - This is necessary so that ruby (and friends) are in the path. We could also have `daemon.rb` set `ruby = File.join(basedir, "sys", "ruby", "bin", "ruby.exe").tr('/', '\\')` so that we're not relying on the search path.
> * `C:\Program Files\Puppet Labs\Puppet\sys\tools\bin` - This contains `elevate` which isn't needed by the daemon, but we should include it for now.

Agreed these values belong in `PATH` - more on that under 4 above.

And note that the implementation here will be different in the Puppet 6 branch.

Reply:
But why bring the puppet 6 changes into puppet 5 in a patch release? I'm thinking the least risky thing is to make sure the environment is configured exactly the same as it was previously in 5.x and 6.x respectively.

Reply:
I'm for just keeping all the ENV variables that were declared in `environment.bat` for each stream, to keep the impact of this changes at minimum . 
About removing `FACTER_env_windows_installdir` i suggest to add the new fact in https://github.com/puppetlabs/facter-ng and remove this variable in puppet 7

Reply:
```Path:  C:\Program Files\Puppet Labs\Puppet\puppet\bin;C:\Program Files\Puppet Labs\Puppet\facter\bin;C:\Program Files\Puppet Labs\Puppet\hiera\bin;C:\Program Files\Puppet Labs\Puppet\mcollective\bin;C:\Program Files\Puppet Labs\Puppet\bin;C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin;C:\Program Files\Puppet Labs\Puppet\sys\tools\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Packer\SysInternals;C:\Program Files\Git\cmd;C:\Program Files\PowerShell\6\;C:\Program Files\Puppet Labs\Puppet\bin;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;
RUBYLIB:  C:/Program Files/Puppet Labs/Puppet/puppet/lib;C:/Program Files/Puppet Labs/Puppet/facter/lib;C:/Program Files/Puppet Labs/Puppet/hiera/lib;C:/Program Files/Puppet Labs/Puppet/mcollective/lib;```